Suggestion - make in_place keyword-only?

First, `polars` is seriously awesome - well done!!

One thing I really like is that you have a `py.typed` file, so types are available to users.

However, the return annotations aren't always as precise as they could be, e.g.:
```python
import polars as pl

df1: pl.DataFrame
df2: pl.DataFrame

reveal_type(df1.vstack(df2))
reveal_type(df1.vstack(df2, in_place=True))
```

If I run this, I get
```
t.py:6: note: Revealed type is "Union[polars.internals.frame.DataFrame, None]"
t.py:7: note: Revealed type is "Union[polars.internals.frame.DataFrame, None]"
```

This could be solved by using `typing.overload`.

**However**, if you have default arguments which precede the one to overload, then you need exponentially many overloads.

In pandas (where changing the API now has to be a slow process!), this has been a nightmare, resulting in constructs like

https://github.com/pandas-dev/pandas/blob/aee6b7ab8a2922606872de083f458d1646dab547/pandas/core/frame.py#L5093-L5206

So, if you'd like to overload return types, I'd strongly suggest making `in_place` (and any other keyword which changes the return type) to be keyword-only while `polars` is still young and the API can change rapidly

---

ps. just started sponsoring you, as I really want `polars` to succeed!

Ok, final question. ;) Could you show me the overload for `DataFrame.sort` with keyword only arguments?

I tried this:

```python
    @tp.overload
    def sort(
        self,
        by: Union[str, "pli.Expr", tp.List[str], tp.List["pli.Expr"]],
        reverse: Union[bool, tp.List[bool]],
        *,
        in_place: Literal[False] = ...,
    ) -> "DataFrame":
        ...

    @tp.overload
    def sort(
        self,
        by: Union[str, "pli.Expr", tp.List[str], tp.List["pli.Expr"]],
        reverse: Union[bool, tp.List[bool]],
        *,
        in_place: Literal[True],
    ) -> None:
        ...

    @tp.overload
    def sort(
        self,
        by: Union[str, "pli.Expr", tp.List[str], tp.List["pli.Expr"]],
        reverse: Union[bool, tp.List[bool]],
        *,
        in_place: bool,
    ) -> Optional["DataFrame"]:
        ...
```

But mypy is not happy.

```python
import polars as pl

df = pl.DataFrame

reveal_type(df.sort(by="a", reverse=False, in_place=True))
```

```
foo.py:4: error: No overload variant of "sort" of "DataFrame" matches argument types "str", "bool", "bool"
foo.py:4: note: Possible overload variants:
foo.py:4: note:     def sort(self, self: DataFrame, by: Union[str, Expr, List[str], List[Expr]], reverse: Union[bool, List[bool]], *, in_place: Literal[False] = ...) -> DataFrame
foo.py:4: note:     def sort(self, self: DataFrame, by: Union[str, Expr, List[str], List[Expr]], reverse: Union[bool, List[bool]], *, in_place: Literal[True]) -> None
foo.py:4: note:     def sort(self, self: DataFrame, by: Union[str, Expr, List[str], List[Expr]], reverse: Union[bool, List[bool]], *, in_place: bool) -> Optional[DataFrame]
foo.py:4: note: Revealed type is "Any"
Found 1 error in 1 file (checked 1 source file)
```

--- a/py-polars/polars/internals/frame.py
+++ b/py-polars/polars/internals/frame.py
@@ -683,10 +683,41 @@ class DataFrame:
         else:
             return {s.name: s.to_list() for s in self}
 
+    @tp.overload
+    def to_json(
+        self,
+        file: Optional[Union[BytesIO, str, Path]] = ...,
+        pretty: bool = ...,
+        *,
+        to_string: Literal[True],
+    ) -> str:
+        ...
+
+    @tp.overload
+    def to_json(
+        self,
+        file: Optional[Union[BytesIO, str, Path]] = ...,
+        pretty: bool = ...,
+        *,
+        to_string: Literal[False] = ...,
+    ) -> None:
+        ...
+
+    @tp.overload
+    def to_json(
+        self,
+        file: Optional[Union[BytesIO, str, Path]] = ...,
+        pretty: bool = ...,
+        *,
+        to_string: bool = ...,
+    ) -> Optional[str]:
+        ...
+
     def to_json(
         self,
         file: Optional[Union[BytesIO, str, Path]] = None,
         pretty: bool = False,
+        *,
         to_string: bool = False,
     ) -> Optional[str]:
         """

